pkg/distributor/distributor.go

```diff
@@ -1093,11 +1093,43 @@ func (d *Distributor) Push(ctx context.Context, req *mimirpb.WriteRequest) (*mim
 		mimirpb.ReuseSlice(req.Timeseries)
 	})
 
-	err := d.PushWithMiddlewares(ctx, pushReq)
-	if err != nil {
-		return nil, err
+	pushErr := d.PushWithMiddlewares(ctx, pushReq)
+	if pushErr == nil {
+		return &mimirpb.WriteResponse{}, nil
 	}
-	return &mimirpb.WriteResponse{}, nil
+	handledErr := d.handlePushError(pushErr)
+	return nil, handledErr
+}
+
+func (d *Distributor) handlePushError(err error) error {
+	if errors.Is(err, context.Canceled) {
+		return err
+	}
+	var distributorError distributor_error.DistributorError
+	if errors.As(err, &distributorError) {
+		switch distributorError := distributorError.(type) {
```

Comment:
There's no need to do `errors.As(err, &distributorError)` if the only thing we doing next is a `switch` on the type and all of them are implementation of it.

Reply:
If we just consider _this_ PR, it is true.
But in the next PR I am planning to add the `ingestererror.Error` interface, that will be used for the errors returned by the ingester. So I was thinking we could distinguish between them and handle them separately.